python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,182 @@
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._fs.touch(self.location)
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Returns a new FsspecInputFile for the location at `self.location`"""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, client_kwargs: Dict[str, str] = EMPTY_DICT, **fs_kwargs):
+        client_properties = {f"client_{key}": val for key, val in client_kwargs.items()}
+        fs_properties = {f"fs_{key}": val for key, val in fs_kwargs.items()}
+        properties = {**client_properties, **fs_properties}
+        self._fs = None
+        super().__init__(properties=properties)
+
+    def new_input(self, location: str) -> FsspecInputFile:
+        """Get an FsspecInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecInputFile: An FsspecInputFile instance for the given location
+        """
+        uri = urlparse(location)
+        fs = self.properties.get("fs") or self._resolve_fs_type(scheme=uri.scheme)
+        return FsspecInputFile(location=location, fs=fs)
+
+    def new_output(self, location: str) -> FsspecOutputFile:
+        """Get an FsspecOutputFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecOutputFile: An FsspecOutputFile instance for the given location
+        """
+        uri = urlparse(location)
+        fs = self.properties.get("fs") or self._resolve_fs_type(scheme=uri.scheme)
+        return FsspecOutputFile(location=location, fs=fs)
+
+    def delete(self, location: Union[str, InputFile, OutputFile]) -> None:
+        """Delete the file at the given location
+
+        Args:
+            location(str, InputFile, OutputFile): The URI to the file--if an InputFile instance or an
+            OutputFile instance is provided, the location attribute for that instance is used as the location
+            to delete
+        """
+        if isinstance(location, (InputFile, OutputFile)):
+            str_location = location.location  # Use InputFile or OutputFile location
+        else:
+            str_location = location
+
+        uri = urlparse(str_location)
+        fs = self._fs or self._resolve_fs_type(scheme=uri.scheme)
+        fs.rm(str_location)
+
+    def _resolve_fs_type(self, scheme: str):
+        """Initialize the proper filesystem for the location's scheme"""
+        if scheme in ("s3", "s3n", "s3a"):
+            from s3fs import S3FileSystem
+
+            self._fs = S3FileSystem(client_kwargs=self._client_properties() or None, **self._fs_properties())
```

Comment:
If we want to leverage the configuration, then we can drop the constructor, and replace this by:
```suggestion
            self._fs = S3FileSystem(client_kwargs=self.properties.get('client_kwargs'), **self.properties.get('fs', {}))
```

Reply:
I think we need to construct `client_kwargs` rather than passing them through. That way we can make configuration the same for both Java and Python FileIO.

Reply:
Updated to explicitly construct client kwargs using the example in your [other comment](https://github.com/apache/iceberg/pull/5332#discussion_r956257603).